libbpf-tools/tcplife.c

```diff
@@ -0,0 +1,255 @@
+// SPDX-License-Identifier: GPL-2.0
+
+/*
+ * tcplife      Trace the lifespan of TCP sessions and summarize.
+ *
+ * Copyright (c) 2022 Hengqi Chen
+ *
+ * Based on tcplife(8) from BCC by Brendan Gregg.
+ * 02-Jun-2022   Hengqi Chen   Created this.
+ */
+#include <argp.h>
+#include <errno.h>
+#include <signal.h>
+#include <time.h>
+#include <arpa/inet.h>
+#include <sys/socket.h>
+
+#include "btf_helpers.h"
+#include "tcplife.h"
+#include "tcplife.skel.h"
+
+#define PERF_BUFFER_PAGES	16
+#define PERF_POLL_TIMEOUT_MS	100
+
+static volatile sig_atomic_t exiting = 0;
+
+static pid_t target_pid = 0;
+static short target_family = 0;
+static char *target_sports = NULL;
+static char *target_dports = NULL;
+static bool emit_timestamp = false;
+static bool wide_column = false;
```

Comment:
What about using `printf("%*d", ...` instead of having a conditional and two different printf lines?

I mean, something like: (I didn't test it). 

```diff
diff --git libbpf-tools/tcplife.c libbpf-tools/tcplife.c
index 411d2c9c..6de4b573 100644
--- libbpf-tools/tcplife.c
+++ libbpf-tools/tcplife.c
@@ -29,7 +29,7 @@ static short target_family = 0;
 static char *target_sports = NULL;
 static char *target_dports = NULL;
 static bool emit_timestamp = false;
-static bool wide_column = false;
+static unsigned int colum_width = 15;
 static bool verbose = false;
 
 const char *argp_program_version = "tcplife 0.1";
@@ -78,7 +78,7 @@ static error_t parse_arg(int key, char *arg, struct argp_state *state)
                target_family = AF_INET6;
                break;
        case 'w':
-               wide_column = true;
+               colum_width = 26;
                break;
        case 'L':
                target_sports = strdup(arg);
@@ -130,15 +130,10 @@ static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
        inet_ntop(e->family, &e->saddr, saddr, sizeof(saddr));
        inet_ntop(e->family, &e->daddr, daddr, sizeof(daddr));
 
-       if (wide_column) {
-               printf("%-7d %-16.16s %-26s %-5d %-26s %-5d %-6.2f %-6.2f %-.2f\n",
-                      e->pid, e->comm, saddr, e->sport, daddr, e->dport,
-                      (double)e->tx_b / 1024, (double)e->rx_b / 1024, (double)e->span_us / 1000);
-       } else {
-               printf("%-7d %-16.16s %-15s %-5d %-15s %-5d %-6.2f %-6.2f %-.2f\n",
-                      e->pid, e->comm, saddr, e->sport, daddr, e->dport,
-                      (double)e->tx_b / 1024, (double)e->rx_b / 1024, (double)e->span_us / 1000);
-       }
+
+       printf("%-7d %-16.16s %-*s %-5d %-*s %-5d %-6.2f %-6.2f %-.2f\n",
+              e->pid, e->comm, colum_width, saddr, e->sport, colum_width, daddr, e->dport,
+              (double)e->tx_b / 1024, (double)e->rx_b / 1024, (double)e->span_us / 1000);
 }
 
 static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
@@ -230,12 +225,9 @@ int main(int argc, char **argv)
 
        if (emit_timestamp)
                printf("%-8s ", "TIME(s)");
-       if (wide_column)
-               printf("%-7s %-16.16s %-26s %-5s %-26s %-5s %-6s %-6s %-s\n",
-                      "PID", "COMM", "LADDR", "LPORT", "RADDR", "RPORT", "TX_KB", "RX_KB", "MS");
-       else
-               printf("%-7s %-16.16s %-15s %-5s %-15s %-5s %-6s %-6s %-s\n",
-                      "PID", "COMM", "LADDR", "LPORT", "RADDR", "RPORT", "TX_KB", "RX_KB", "MS");
+
+       printf("%-7s %-16.16s %-*s %-5s %-*s %-5s %-6s %-6s %-s\n",
+              "PID", "COMM", colum_width, "LADDR", "LPORT", colum_width, "RADDR", "RPORT", "TX_KB", "RX_KB", "MS");
 
        while (!exiting) {
                err = perf_buffer__poll(pb, PERF_POLL_TIMEOUT_MS);
```

Reply:
Thanks, I didn't know this before.